### DISCUSSION TITLE
Any alternative for Router events in V13.4

### DISCUSSION THREAD

**Discussion by market-south** [vercel/next.js#51329] [App Router]
Hi there, 

I am working on a project which has a mega menu. And using a `useState` to handle the menu open and close. We used `Router.events.on` and `Router.events.off` in V12 for handling the useState when the user navigates between pages. 

I understand V13 doesn't support Router events anymore.  After doing some research, I created a workaround to handle the useState when the user navigates between pages. But unable to hide the menu when the user navigates to the same page. 

Example code as below:

```
 const [megaMenu, setMegaMenu] = useState(false);

{/* TOGGLE THE STATE *}
    const toggleMenuBar = () => {
        setMegaMenu((b) => !b);
    };

{/* HAMBURGER ICON TOGGLE */}
    const toggleMenu = () => {
        if (megaMenu) return styles["active"];
        else return "";
    };

{/* TOGGLE MEGA MENU */}
    const toggleMenuItem = () => {
        if (megaMenu) return styles["active"];
        else return "";
    };

{/* CLOSE THE MENU *}
    const hideMegaMenu = (url) => {
        setMegaMenu(false);
    };

    const currentPage = usePathname();  // Get current route
    const savedPathNameRef = useRef(currentPage);

    useEffect(() => {

        if (savedPathNameRef.current !== currentPage) {
            hideMegaMenu();
            savedPathNameRef.current = currentPage;
        }

    }, [currentPage, hideMegaMenu]);
```

This works perfectly when the user uses hamburger to toggle the menu or when the menu is open and tries to navigate to another page. But unable to close the menu when the user clicks the same page link in the menu. In V12 we used `RouteChangeComplete` to close the menu if the user navigates to the same page.

Any help would be appreciated.

**Comment by icyJoseph** [comment]
Try with this hook: https://github.com/vercel/next.js/discussions/41745#discussioncomment-3987025
